### PR TITLE
Re-write of the sign up logic + include sign up wizard on navbar

### PR DESCRIPTION
**Pull request by johancc**

### CODE REVIEW

File: packages/client/src/api.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        await Auth.createUserWithEmailAndPassword(email, password);
+        token = await Auth.currentUser.getIdToken();
+        await post('/users', { role: role, ...data}, { token });
+        await Auth.currentUser.sendEmailVerification();
+        
+    } catch (err) {
+        if (token) await Auth.currentUser.delete();
+        throw new Error(`Error creating ${role}: ${err}`);
+    }
+}
```
Comment (sanjayyepuri) [inline packages/client/src/api.js:44]: nit: add new line.

File: packages/client/src/pages/FindAMentor/MentorSearch.css [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+.hit-picture img {
+  width: 80px;
+  height: 80px;
+}
+
+.hit-type {
+  color: #888888;
+  font-size: 13px;
+}
```
Comment (sanjayyepuri) [inline packages/client/src/pages/FindAMentor/MentorSearch.css:95]: nit: new line

File: packages/server/db/user.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
+const firebase = require('firebase-admin');
```
Comment (sanjayyepuri) [inline packages/server/db/user.js:1]: This file seems to be overwriting the changes in master. Is that intended?

File: packages/server/middleware/auth.js [hunk 4]
```diff
@@ -2,8 +2,9 @@ const firebase = require('firebase-admin');
 
 async function authMiddleware(req, res, next) {
   try {
+    // The token should only ever be sent in a header.
     req.user = await firebase.auth()
-      .verifyIdToken(req.query.token || req.headers.token || req.body.token);
```
Comment (sanjayyepuri) [inline packages/server/middleware/auth.js:6]: Can we just remove req.query? I think it is fine for us to be getting the token from the body.

File: packages/server/app.js [hunk 5]
```diff
@@ -4,14 +4,11 @@ const path = require('path');
...
 
-const app = express();
+// Firebase boilerplate.
+const firebase = require('firebase-admin');
 
-// disable when running unit tests
-if (process.env.NODE_ENV !== 'test') {
-  app.use(logger('dev'));
-}
+const serviceAccount = require('./service_account.json');
```
Comment (sanjayyepuri) [inline packages/server/app.js:11]: If we import firebase credentials using require, we don't need line 16. The require call will fail if the file does not exist.

File: packages/client/src/utilities.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          throw Error(`API request's result could not be converted to a JSON object: \n${text}`);
+        });
+      });
+  }
+
+  // Helper code to make a get request. Default parameter of empty JSON Object for params.
+  // Returns a Promise to a JSON Object.
+  export function get(endpoint, params = {}, headers = {}) {
+    const fullPath = endpoint + '?' + formatParams(params);
+    return fetch(fullPath, { headers })
```
Comment (sanjayyepuri) [inline packages/client/src/utilities.js:43]: Perfect ;)

File: packages/client/src/pages/SignUp/wizards/ParentWizard/index.js [hunk 7]
```diff
@@ -123,7 +152,7 @@ const ThirdPage = () => {
             {children}
             <ChildSignUpButtonWrapper>
                 <Button onClick={handleAddClick}>
-                    Add Child
+                    Add Another Child
```
Comment (sanjayyepuri) [inline packages/client/src/pages/SignUp/wizards/ParentWizard/index.js:155]: nit: Drop the "Another."

File: packages/client/src/pages/SignUp/wizards/ParentWizard/index.js [hunk 8]
```diff
@@ -61,6 +63,33 @@ const updateRegisteredChild = (index, data) => {
...
+        name: parentData.parentName,
+        phone: parentData.parentPhoneNumber,
+        timezone: parentData.timeZone,
+        email: parentData.parentEmail,
+        students: parentData.registeredChildren
+            .map((studentData) => createStudentModel(studentData)),
+        role: 'PARENT',
+    }
+}
+
```
Comment (sanjayyepuri) [inline packages/client/src/pages/SignUp/wizards/ParentWizard/index.js:92]: We should move these functions to a shared file. They may be useful when building the Profile page.

File: packages/client/src/pages/SignUp/wizards/MentorWizard/index.js [hunk 9]
```diff
@@ -39,6 +40,15 @@ let mentorWizardSignUpData = {
...
 
+const createMentorModel = (mentorData) => ({
+    name: mentorData.mentorName,
+    email: mentorData.mentorEmail,
+    gradeLevels: mentorData.selectedGradeLevels,
+    subjects: mentorData.selectedSubjects,
+    major: mentorData.major,
+    role: 'MENTOR', 
+    introduction: mentorData.introduction,
+})
```
Comment (sanjayyepuri) [inline packages/client/src/pages/SignUp/wizards/MentorWizard/index.js:51]: Let's move these createXXXModel functions to a shared file.

File: packages/client/src/pages/FindAMentor/MentorSearch.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                      <Pagination showLast={true} />
+                    </div>
+                </div>
+            </InstantSearch>
+    </div>
+    
+);
+
+
+export default MentorSearch;
```
Comment (sanjayyepuri) [inline packages/client/src/pages/FindAMentor/MentorSearch.js:52]: nit: new line

File: packages/client/src/pages/FindAMentor/MentorRequestFrame.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import MentorProfile from './MentorProfile';
+
+// Shows more detailed mentor information and allows parents to request mentors.
+const MentorRequestFrame = ({ mentor }) => (<MentorProfile mentor={mentor}/>);
+
+export default MentorRequestFrame;
```
Comment (sanjayyepuri) [inline packages/client/src/pages/FindAMentor/MentorRequestFrame.js:7]: nit: new line

File: packages/client/src/pages/FindAMentor/MentorGrid.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                    <Modal title={mentor.name} key={mentor._id} trigger={<div><MentorCard mentor={mentor} /></div>}>
+                        <MentorRequestFrame mentor={mentor}/>
+                    </Modal>
+                );
+            })}
+        </MentorGridContainer>
+    )
+}
+
+export default MentorGrid;
```
Comment (sanjayyepuri) [inline packages/client/src/pages/FindAMentor/MentorGrid.js:88]: nit: new line

File: packages/client/src/pages/FindAMentor/MentorSearch.css [hunk 13]
```diff
@@ -0,0 +1,95 @@
+.ais-SearchBox {
```
Comment (sanjayyepuri) [inline packages/client/src/pages/FindAMentor/MentorSearch.css:1]: Are we doing css files?

File: packages/client/src/components/SignIn/index.js [hunk 14]
```diff
@@ -117,13 +117,8 @@ const Signin = () => {
...
         if (formError || serverError) return;
         signin(email, password)
-            .then((resp) => {
-                // Redirect.
-                alert('Success! Redirecting to sign in...');
-            })
-            .catch(() => {
-                setServerError(true);
-            });
+            .then(() => console.log('Signed in!'))
```
Comment (sanjayyepuri) [inline packages/client/src/components/SignIn/index.js:120]: Don't print to console.

**Review by sanjayyepuri — CHANGES_REQUESTED** [review]

File: packages/server/middleware/auth.js (hunk 4 above)
Comment (johancc) [reply packages/server/middleware/auth.js:6]: I think that GET request will put the token in the query and we want to make sure the middleware handles all the cases. Ideally, all requests should put it in the header but I don't think our code abides by that

File: packages/client/src/pages/FindAMentor/MentorSearch.css (hunk 13 above)
Comment (johancc) [reply packages/client/src/pages/FindAMentor/MentorSearch.css:1]: @oupadhyay will remove this an upcoming CL

**Comment by johancc** [comment]
Leaving this to @oupadhyay

**Pull request merged**